### PR TITLE
marshal chaincode return message as payload to calling chaincode

### LINKED ISSUE #1425: In CC to CC calls, the calling CC will get killed by FSM if called CC returns Error.
## Description

If you set up a chaincode (cc) that calls another cc, then if the called cc returns a fmt.Error (which can be common), then the cc FSM intercepts that Error and kills the calling cc.  This does not happen if you call the cc from a client (ie outside the chain).  The behavior should be consistent across calling methods (ie should not depend on who is calling the cc) and the common use case of returning a fmt.Errror should be supported.
## Describe How to Reproduce

Attached are two chaincodes: dns.go and dns-test.go that demonstrate this problem.  To reproduce follow these steps:
- Run cc dns.go and dns-test.go on the peer in dev mode in terms 1 and 2.
  - term1: `CORE_CHAINCODE_ID_NAME=dns CORE_PEER_ADDRESS=0.0.0.0:30303 ./dns`
  - term2: `CORE_CHAINCODE_ID_NAME=dns-test CORE_PEER_ADDRESS=0.0.0.0:30303 ./dns-test`
- Deploy both cc's from term3.
  - `./peer chaincode deploy -n dns -c '{"Function":"FooBar", "Args":[]}'`
  - `./peer chaincode deploy -n dns-test -c '{"Function":"FooBar", "Args":[]}'`
- Create a namespace in term3
  - `./peer chaincode invoke -l golang -n dns-test -c '{"Function": "registerNamespace", "Args":["NS1", "owner1"]}'`
- Create a name mapping in term3
  - `./peer chaincode invoke -l golang -n dns-test -c '{"Function": "register", "Args":["owner1", "NS1", "test", "1234"]}'`
- Call query for existing name (should not fail)
  - `./peer chaincode query -l golang -n dns-test -c '{"Function": "lookup", "Args":["NS1", "test"]}'`
- Call query for non-existing name (will fail)
  - `./peer chaincode query -l golang -n dns-test -c '{"Function": "lookup", "Args":["NS1", "HelloWorld"]}'`

The error seen in term2 (which is running dns-test) is:

`Query: (lookup)
Calling (dns), function (lookup) with args ([NS1 HelloWorld])
2016/05/05 22:24:13 [683d6d42]Sending INVOKE_QUERY
2016/05/05 22:24:13 [683d6d42]Received message ERROR from shim
2016/05/05 22:24:13 [683d6d42]Handling ChaincodeMessage of type: ERROR(state:ready)
Error starting dns-test: Error handling message: [683d6d42-5912-4a28-9d9e-4b8475586b46]Chaincode handler FSM cannot handle message (ERROR) with payload size (31) while in state: ready`

And the dns-test cc process exits as a result.

### PR DESCRIPTION
**Pull request by muralisrini**
Chaincode calling chaincode treated all errors - fabric as well as error returned from called chaincode - as errors. This caused calling chaincode to error out. 
## Description

chaincode_support.go - Chaincode's execute method returns response and error. It used to introspect response and if a chaincode error,  set the returned error as well. This is unnecessary as the fabric treats chaincode errors correctly anyway. By not setting the error in the first place the way is paved for chaincode-calling-chaincode path. 

With Execute not returning error for everything, rest of the fix can rely upon nil error to marshalls the response from the called chaincode in the peer and unmarshalls it in the calling shim. The response is then checked for error in calling chaincode.
## Motivation and Context

Conflating fabric errors with chaincode errors causes these to be indistiguishable in the called chaincode. Called chaincode treats all errors as Fabric errors and exit with error.  The fix disambiguates the two cases and the shim correctly  looks for fabric errors, chaincode errors and chaincode success responses.

Fixes #1425
## How Has This Been Tested?

Tested with testcase supplied in #1425 
Added unit tests
## Checklist:

- [X] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [X] Either no new documentation is required by this change, OR I added new documentation
- [X] Either no new tests are required by this change, OR I added new tests
- [X] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: muralisr@us.ibm.com

### CODE REVIEW

**Comment by ghaskins** [comment]
Thank you @muralisrini 

LGTM

Reviewed-by: Greg Haskins gregory.haskins@gmail.com

File: orderer/common/filerepo/filerepo.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+	if err := w.Flush(); err != nil {
+		return err
+	}
+
+	if err := tmpFile.Close(); err != nil {
+		return err
+	}
+
+	return os.Rename(tmpFile.Name(), dest)
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: **Close and rename do not imply sync** -- a good deck that exactly talk about this scenario https://www.slideshare.net/nan1nan1/eat-my-data

As you might know, `Flush` flushes the data present in the write buffer only to the disk (no fsync).

File: orderer/common/filerepo/filerepo.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+// Scan parses the directory and produce a list of file names, filtered by suffix.
+func (r *Repo) Scan() ([]string, error) {
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: Don't we need synchronization between `Scan()`, `Read()`, and `Save()` using a lock? If not, can you clarify how these operations can interleave and still provide consistency?

File: orderer/common/filerepo/filerepo.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	"bufio"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hyperledger/fabric/common/configtx"
+)
+
+type Repo struct {
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go:29]: Don't we need a `Remove()` method on the Repo? I assume that it is missing because the consumer of this pkg does not need it.

File: orderer/common/filerepo/filerepo.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	Options Options
+}
+
+type Options struct {
+	FileSuffix          string
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: I assume the file suffix as a type of entry stored?

File: orderer/common/filerepo/filerepo.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	TransientFilePrefix string
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: Isn't this field internal to this pkg? Should we need to export it?

File: orderer/common/filerepo/filerepo.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+	tmpFile, err := ioutil.TempFile(fileRepoDir, r.Options.TransientFilePrefix+channelName)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err != nil {
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: It doesn't guarantee a removal right? Given that there is an error already, I am not sure whether the removal/sync would even succeed.

File: orderer/common/filerepo/filerepo.go (hunk 3 above)
Comment (tock-ibm) [reply orderer/common/filerepo/filerepo.go:29]: That is correct, we need `Remove` when on boarding is done. I think it should be added.

File: orderer/common/filerepo/filerepo.go (hunk 5 above)
Comment (tock-ibm) [reply orderer/common/filerepo/filerepo.go]: I think - leave it exposed but apply a default `~` for when it is not set.

File: orderer/common/filerepo/filerepo.go (hunk 2 above)
Comment (tock-ibm) [reply orderer/common/filerepo/filerepo.go]: Thread safety is not just thread races - concurrent access - but also visibility. In the channel participation API, the `Scan` is used by the main go-routine, when initializing the server, `Save` by the http server request go-routine (which probably may spin multiple go-routines if there are multiple requests?), and the `Remove` by the internal go-routine servicing the follower.Chain, one for each chain. Since this is used by multiple threads, I think a thread safe implementation is needed.

File: orderer/common/filerepo/filerepo.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	FileSuffix          string
+	TransientFilePrefix string
+	FileRepoDir         string
+}
+
+// Save atomically persists the content to the channelName+suffix file by first writing it
+// to a tmp file prefixed by the TransientFilePrefix and then moves the file to the final
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(channelName string, content []byte) error {
+	err := configtx.ValidateChannelID(channelName)
```
Comment (tock-ibm) [inline orderer/common/filerepo/filerepo.go]: I don't think we need to validate the channel name... channel names are very strict, and this would limit the generality of this package. W.r.t the channel participation API, we check the validity of the channel name when we first process the request, so no need to do it here.

File: orderer/common/filerepo/filerepo.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		return err
+	}
+
+	fileRepoDir := r.Options.FileRepoDir
+	dest := filepath.Join(fileRepoDir, channelName+r.Options.FileSuffix)
+
+	tmpFile, err := ioutil.TempFile(fileRepoDir, r.Options.TransientFilePrefix+channelName)
```
Comment (tock-ibm) [inline orderer/common/filerepo/filerepo.go]: I have second thoughts on this... maybe it would be much cleaner to add the transient marker `~` as appended to the  suffix? like:
`mychannel.joinblock` vs. `mychannel.joinblock~` 
This seems to be closer to what many other programs do, and, this way there will never be a confusion with object names (i.e. channel names) starting with a `~`?

File: orderer/common/filerepo/filerepo.go (hunk 1 above)
Comment (tock-ibm) [reply orderer/common/filerepo/filerepo.go]: You have to use `Sync()` on the file, I think, to write both data and metadata to disk.

File: orderer/common/filerepo/filerepo.go (hunk 8 above)
Comment (tock-ibm) [reply orderer/common/filerepo/filerepo.go]: Plus, why a temp file? as we have the transient marker?

File: orderer/common/filerepo/filerepo.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+// to a tmp file prefixed by the TransientFilePrefix and then moves the file to the final
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(channelName string, content []byte) error {
+	err := configtx.ValidateChannelID(channelName)
+	if err != nil {
+		return err
+	}
+
+	fileRepoDir := r.Options.FileRepoDir
+	dest := filepath.Join(fileRepoDir, channelName+r.Options.FileSuffix)
```
Comment (tock-ibm) [inline orderer/common/filerepo/filerepo.go]: I think we want to fail early if `dest` above already exists, otherwise the os.Rename() will fail us later.

File: orderer/common/filerepo/filerepo.go (hunk 6 above)
Comment (tock-ibm) [inline orderer/common/filerepo/filerepo.go]: This may leave transient files on the dir after crash. When we start using the repo again, after a crash, we should clean all the transient files. In the participation API we use `Scan` first, so that might be a good place to do it (once!). However, I think a better way is to (optionally) make `options` private, and build a constructor that cleans transient files. This way we force the user towards a safe and correct usage pattern.

```go
type Repo struct {
  options Options
}

func NewRepo( options Options) (*Repo, error) {
  //Create the directory if not there
  //Scan and remove transient files
  //Build a Repo and return it
}
```

File: orderer/common/filerepo/filerepo.go (hunk 3 above)
Comment (tock-ibm) [inline orderer/common/filerepo/filerepo.go:29]: I would consider being more specific about the file permissions:

```go
	repoFilePermPrivateRW os.FileMode = 0600
	repoDirPermPrivateRWX os.FileMode = 0700
```

In addition, consider creating a separate directory for every suffix, under the parent dir. This way the `*.joinblock` and `*.remove` files will be in different sub-directories. This would speed up the scan and avoid filesystem sync issues if we use two file-repo's in the same process with different suffixes. 
Another way to do this is to explicitly define a different file repo dir for different uses in the orderer.yaml, for example.

File: orderer/common/filerepo/filerepo_test.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/orderer/common/filerepo"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestFileRepo_Save(t *testing.T) {
```
Comment (tock-ibm) [inline orderer/common/filerepo/filerepo_test.go]: Add a test for saving and loading empty files with a `nil` or empty []byte content, for the `channel.remove` use case.

File: orderer/common/filerepo/filerepo_test.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+		assert.True(t, os.IsNotExist(err))
+	})
+
+	t.Run("Failed save, invalid channel name", func(t *testing.T) {
+		err := r.Save("InvalidChannelName", []byte{})
+		assert.EqualError(t, err, "'InvalidChannelName' contains illegal characters")
+	})
+}
+
+func TestFileRepo_Scan(t *testing.T) {
```
Comment (tock-ibm) [inline orderer/common/filerepo/filerepo_test.go]: In the context of the above discussion about cleaning temp files after a crash, add a test case that puts some dummy temp files in the directory and test that the first `Scan` or `NewRepo` cleans them up (I prefer the latter).

File: orderer/common/filerepo/filerepo.go [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+	// Remove existing transient files in the repo
+	for _, f := range files {
+		transientFilePattern := "*" + repo.fileSuffix + repo.transientFileMarker
+		isTransientFile, err := filepath.Match(transientFilePattern, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isTransientFile {
+			os.Remove(filepath.Join(repo.fileRepoDir, f.Name()))
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: We need to call `sync()` on the parent directory.

File: orderer/common/filerepo/filerepo.go [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+func New(repoParentDir, fileSuffix string, opts ...RepoOption) (*Repo, error) {
+	// default transientFileMarker to ~, can modify via opts
+	repo := &Repo{
+		transientFileMarker: "~",
+		fileSuffix:          fileSuffix,
+		fileRepoDir:         filepath.Join(repoParentDir, fileSuffix),
+	}
+
+	for _, opt := range opts {
+		if err := opt(repo); err != nil {
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: additional response to @sykesm question on why I find this hard to read. 

1. I did go to definition by keeping the cursor on `opt`. Nothing happened.
2. I needed to track to opts and reach `ReadOption`
3. Then, I find that ReadOption is a function pointer and it provides a list of functions which the consumer could have called to get a function implementation that sets a value for a variable. 
4. The returned function implementation is called here. 

I understand that this allow the caller to configure only the required variables (i.e., optional variables) but adds more time to the code readability (& multi-level functions). As I do not see any strong reason for the caller to set `transientFileMarker` (as it is internal to the `filerepo` pkg), I am not convinced that this added code complexity is required. The default `~` looks fine to me. If others have a strong opinion that the added code complexity is justified and `transientFileMarker` needs to be configured by the caller, I will step aside.

File: orderer/common/filerepo/filerepo.go [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+// All file system operations on the returned file repo are thread safe.
+func New(repoParentDir, fileSuffix string) (*Repo, error) {
+	// Check that fileSuffix doesn't contain the os path separator
+	if err := validateFileSuffix(fileSuffix); err != nil {
+		return nil, err
+	}
+
+	fileRepoDir := filepath.Join(repoParentDir, fileSuffix)
+
+	// Make the file repo if it doesn't already exist
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: Nit. Would be good to have comments to explain a decision or reasoning or complex side-effects. Other types of comments are distraction while reading the code.

File: orderer/common/filerepo/filerepo.go [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+	"github.com/hyperledger/fabric/internal/fileutil"
+	"github.com/pkg/errors"
+)
+
+const (
+	repoFilePermPrivateRW os.FileMode = 0600
+)
+
+type Repo struct {
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go:29]: Nit. Doc comment for this exported type is missing.

File: orderer/common/filerepo/filerepo.go [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+	// Remove existing transient files in the repo
+	for _, f := range files {
+		transientFilePattern := "*" + fileSuffix + "~"
+		isTransientFile, err := filepath.Match(transientFilePattern, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isTransientFile {
+			os.Remove(filepath.Join(fileRepoDir, f.Name()))
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: Better to check the err returned by `os.Remove()`

File: orderer/common/filerepo/filerepo.go [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+// FileToBaseName strips the suffix from the file name to get the associated channel name.
+func (r *Repo) FileToBaseName(fileName string) string {
+	baseFile := filepath.Base(fileName)
+
+	return strings.TrimSuffix(baseFile, "."+r.fileSuffix)
+}
+
+func validateFileSuffix(fileSuffix string) error {
+	if len(fileSuffix) <= 0 {
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: Nit. as it can never be < 0
```suggestion
	if len(fileSuffix) == 0 {
```

File: orderer/common/filerepo/filerepo.go [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+	tmpFileName := fileName + r.transientFileMarker
+	if err := fileutil.CreateAndSyncFileAtomically(r.fileRepoDir, tmpFileName, fileName, content, repoFilePermPrivateRW); err != nil {
+		return err
+	}
+
+	return fileutil.SyncDir(r.fileRepoDir)
+}
+
+// Remove removes the file associated with baseName from the file system.
+func (r *Repo) Remove(baseName string) error {
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go:106]: Unit test is missing.

File: orderer/common/filerepo/filerepo.go [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	files, err := ioutil.ReadDir(fileRepoDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove existing transient files in the repo
+	for _, f := range files {
+		transientFilePattern := "*" + fileSuffix + "~"
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: Nit. Can be kept outside the loop instead of creating it newly for each file.

Comment (cendhu) [reply orderer/common/filerepo/filerepo.go]: Nit. `~` -- till we reach the line 74, the reader cannot understand what this mean. One way to fix this is by using a const variable to keep the default transientFileMarker.

File: orderer/common/filerepo/filerepo.go [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+
+// Save atomically persists the content to suffix/baseName+suffix file by first writing it
+// to a tmp file marked by the transientFileMarker and then moves the file to the final
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(baseName string, content []byte) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	fileName := fmt.Sprintf("%s.%s", baseName, r.fileSuffix)
+	dest := filepath.Join(r.fileRepoDir, fileName)
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go]: Nit. Given that these two lines occurs in three methods on `Repo`, it would be good to move them to a separate method called `baseToFileName()` (which would do an opposite of the existing `FileToBaseName()`)

File: orderer/common/filerepo/filerepo.go [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+
+// New initializes a new file repo at repoParentDir/fileSuffix.
+// All file system operations on the returned file repo are thread safe.
+func New(repoParentDir, fileSuffix string) (*Repo, error) {
+	// Check that fileSuffix doesn't contain the os path separator
+	if err := validateFileSuffix(fileSuffix); err != nil {
+		return nil, err
+	}
+
+	fileRepoDir := filepath.Join(repoParentDir, fileSuffix)
```
Comment (cendhu) [inline orderer/common/filerepo/filerepo.go:43]: Nit. Using `fileSuffix` both as a sub-directory name and also as a suffix to the file name within the sub directory kind of confuses me while reading the code. I think the fileSuffix denotes the type of content being stored. One such type is the join block. We then create a sub-folder for the type and also add the type as a suffix to each file stored in the sub-folder.

Anyway, I do not have an alternative to the current folder structure that is shown below. Maybe the joinblock suffix looks redundant. 
```
/repo/joinblock
/repo/joinblock/channel1.joinblock
/repo/joinblock/channel2.joinblock
/repo/joinblock/channel3.joinblock~
```
